src/main/resources/templates/product/productDetails.html

```diff
@@ -125,6 +125,66 @@
                     <button class="mdl-chip__action" type="button"><i class="material-icons">info</i></button>
                 </span>
         </div>
+        <!-- review button -->
+        <div sec:authorize="hasRole('USER')" class="add-review">
+            <a th:if="${review.reviewText} == null" class="add-review-link" onclick="showReviewForm()">
+                <label style="cursor: pointer;" class="add-review-text">Add Your Review</label>
+                <em class="material-icons">chat</em>
+            </a>
+        </div>
+        <!-- review form -->
+        <div class="review-form" style="display: none" sec:authorize="hasRole('USER')">
+            <form method="post" th:object="${review}"
+                  th:action="${review.reviewText} == null ?
+                  @{/products/details/{id}/addreview(id=${product.id})} :
+                  @{/products/details/{id}/updatereview(id=${product.id})}">
+                <input type="hidden" id="userName" th:field="*{userName}">
+                <div class="mdl-textfield mdl-js-textfield mdl-textfield--floating-label">
+                    <textarea rows="4" class="mdl-textfield__input" type="text" id="reviewText"
+                              th:field="*{reviewText}"></textarea>
+                    <label class="mdl-textfield__label" for="reviewText">Review</label>
+                </div>
+                <div class="review-form-submit-div">
+                    <button class="mdl-button--colored review-form-submit" type="submit">
+                        <a title="Submit review" class="material-icons review-form-submit-icon">check</a>
+                    </button>
+                    <a th:href="@{/products/details/{id}(id=${product.id})}"
+                       title="Submit review" class="material-icons review-form-cancel-icon">close</a>
+                </div>
+            </form>
+        </div>
+        <!-- review list -->
+        <div class="review-list" th:if="${null != product.reviews}">
+            <ul style="list-style: none;">
+                <li class="review-list-review-item" th:each="review : ${product.reviews}">
+                    <div class="review-list-review-item-talkbubble"
+                         th:classappend="${#authorization.expression('hasRole(''ADMIN'')')} OR ${review.userName} == ${loggedInUser} ? 'review-list-review-item-talkbubble-active'">
```

Comment:
you don't need an attribute for loggedInUser -> "${#authentication.getName()}"
OR reads like an SQL statement, could be | |

Reply:
changed authentication, OR is needed, || is problematic , i started with || when making it but had to switch to OR

Reply:
damn thymeleaf